### ISSUE TITLE
does not support bf16-only training

### ISSUE THREAD

**Issue by bghira** [zyushun/Adam-mini#26]
hello, great work. however, this as currently stands requires autocast for training in bf16 so that we use fp32 gradients during updates?

i was hoping you could investigate adding kahan summation and stochastic rounding to allow this use without the model degrading into its rounding bias.

**Comment by zyushun** [comment]
Hi @bghira ! Thanks a lot for mentioning this.

fp32 gradient is not necessarily needed. We have refactored the code and we now remove the requirement of fp32 from the lines “grad = p.grad.to(torch.float32)” and  " state["m"] = torch.zeros_like(p, dtype = torch.float32)" . See below as an example.

```
    @torch.no_grad()
    def step(self, closure=None):
        loss = None
        if closure is not None:
            with torch.enable_grad():
                loss = closure()
        for group in self.param_groups:
            beta1 = group["beta1"]
            beta2 = group["beta2"]
            lr = group["lr"]
            name = group["name"]
            eps = group["eps"]

            for p in group["params"]:
                
                state = self.state[p]
                if any(embd_name in name for embd_name in self.embd_names):  # this is for embedding and output layer
                    if p.grad is None:
                        continue
                    if len(state) == 0:
                        state["m"] = torch.zeros_like(p, memory_format=torch.preserve_format) # dtype = torch.float32)
                        state["step"] = 0
                        state["v"] = torch.zeros_like(p, memory_format=torch.preserve_format) # dtype = torch.float32)

                    grad = p.grad #.to(torch.float32)
```


We will update  the code in pip as soon as we've done some additional testing. I will keep you noticed  when we do.

Regarding your suggestions on "adding kahan summation and stochastic rounding to allow this use without the model degrading into its rounding bias." We apologize that we are not sure how to do this and perhaps we need some preliminary knowledge to implement this. We are very willing to learn and please feel free to share some sample code if there is any.
